File: apps/desktop/src/main/signaling/handshake.test.ts
import { describe, expect, it } from 'vitest'
import { createNonce, computeMac, verifyMac, negotiateProtocol, NONCE_BYTES } from './handshake'
import type { HelloPayload } from './envelope'

// Cross-platform interop vector: the identical values and expected mac are
// asserted in the Android unit test (apps/mobile …/signaling/HandshakeTest.kt),
// proving both HMAC implementations agree byte-for-byte.
const VECTOR_SECRET_B64URL = 'AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8'
const VECTOR_SID = 'Jm1LIOO4mWm0lRSSl2fClw'
const VECTOR_NONCE = 'EBESExQVFhcYGRobHB0eHw'
const VECTOR_MAC = 'cPTgkuJCjSp2z9MZyxVs-QtHs8fCYrMN7s-k0YfyXZQ'

describe('handshake crypto (pairing.md / ADR-002)', () => {
  it('matches the cross-platform HMAC-SHA256 vector', () => {
    const secret = Buffer.from(VECTOR_SECRET_B64URL, 'base64url')
    const mac = computeMac(secret, VECTOR_SID, VECTOR_NONCE)
    expect(mac).toBe(VECTOR_MAC)
    expect(verifyMac(VECTOR_MAC, mac)).toBe(true)
  })

  it('rejects a wrong mac, a wrong nonce and a wrong secret', () => {
    const secret = Buffer.from(VECTOR_SECRET_B64URL, 'base64url')
    const good = computeMac(secret, VECTOR_SID, VECTOR_NONCE)

    expect(verifyMac(good, computeMac(secret, VECTOR_SID, 'different-nonce'))).toBe(false)
    expect(verifyMac(good, computeMac(Buffer.from('x'.repeat(32), 'ascii'), VECTOR_SID, VECTOR_NONCE))).toBe(false)
    expect(verifyMac(good, good.slice(0, 42))).toBe(false)
  })

  it('issues a fresh base64url 16-byte nonce per call', () => {
    const first = createNonce()
    const second = createNonce()
    expect(Buffer.from(first, 'base64url')).toHaveLength(NONCE_BYTES)
    expect(first).toMatch(/^[A-Za-z0-9_-]+$/)
    expect(first).not.toBe(second)
  })

  it('negotiates the highest mutually supported protocol version', () => {
    const hello = (protoMin: number, protoMax: number): HelloPayload => ({ ua: 'test', protoMin, protoMax })
    expect(negotiateProtocol(hello(1, 1))).toBe(1)
    expect(negotiateProtocol(hello(1, 9))).toBe(1)
    expect(negotiateProtocol(hello(2, 9))).toBeNull()
    expect(negotiateProtocol(hello(0, 0))).toBeNull()
  })
})

File: apps/mobile/app/src/test/java/com/zerofriction/localcast/signaling/HandshakeTest.kt
package com.zerofriction.localcast.signaling

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Test

class HandshakeTest {

    @Test
    fun `matches the cross-platform HMAC-SHA256 vector`() {
        // Identical values and expected mac are asserted in the desktop vitest
        // (apps/desktop/src/main/signaling/handshake.test.ts) — both platforms
        // must agree byte-for-byte.
        val secret = Handshake.decodeSecret(VECTOR_SECRET_B64URL)

        assertArrayEquals(bytes(0, 32), secret)
        val mac = Handshake.computeMac(secret, VECTOR_SID, VECTOR_NONCE)

        assertEquals(VECTOR_MAC, mac)
    }

    @Test
    fun `a different nonce produces a different mac`() {
        val secret = Handshake.decodeSecret(VECTOR_SECRET_B64URL)
        val mac = Handshake.computeMac(secret, VECTOR_SID, VECTOR_NONCE)

        assertFalse(mac == Handshake.computeMac(secret, VECTOR_SID, "EBESExQVFhcYGRobHB0eAA"))
        assertFalse(mac == Handshake.computeMac(secret, "AnotherSessionId", VECTOR_NONCE))
    }

    private fun bytes(from: Int, until: Int): ByteArray = (from until until).map { it.toByte() }.toByteArray()

    private companion object {
        const val VECTOR_SECRET_B64URL = "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8"
        const val VECTOR_SID = "Jm1LIOO4mWm0lRSSl2fClw"
        const val VECTOR_NONCE = "EBESExQVFhcYGRobHB0eHw"
        const val VECTOR_MAC = "cPTgkuJCjSp2z9MZyxVs-QtHs8fCYrMN7s-k0YfyXZQ"
    }
}
